### PR TITLE
Make `buildQuorumConfig` to use UTXO instead of PublicKey

### PR DESCRIPTION
**Pull request by linked0**
This change helps use UTXOs in NameRegistry. `Validator` makes a set of UTXO for peers, which will be used to register peers in NetworkManager.

Relates to #1309

### CODE REVIEW

File: source/agora/node/Validator.d [hunk 1]
```diff
@@ -200,6 +205,13 @@ public class Validator : FullNode, API
             log.fatal("rebuildQuorumConfig: Exception thrown: {}", e);
             assert(0);
         }
+
+        foreach (hash; keys.filter!(key => key != this_enroll).array)
```
Comment (hewison-chris) [inline source/agora/node/Validator.d]: Is the `.array` required?

**Review by hewison-chris — APPROVED** [review]
LGTM

File: source/agora/node/Validator.d (hunk 1 above)
Comment (linked0) [reply source/agora/node/Validator.d]: I just realized that it isn't. Will remove.

File: source/agora/node/Validator.d [hunk 2]
```diff
@@ -200,6 +205,13 @@ public class Validator : FullNode, API
             log.fatal("rebuildQuorumConfig: Exception thrown: {}", e);
             assert(0);
         }
+
+        foreach (hash; keys.filter!(key => key != this_enroll))
+        {
+            UTXO utxo;
+            assert(this.utxo_set.peekUTXO(hash, utxo));
+            this.required_peer_utxos[hash] = utxo;
+        }
```
Comment (Geod24) [inline source/agora/node/Validator.d]: This should probably be in `buildRequiredKeys` instead (and the removed call be reinstated).

File: source/agora/node/Validator.d [hunk 3]
```diff
@@ -140,8 +140,6 @@ public class Validator : FullNode, API
         static QuorumConfig[] other_qcs;
         this.rebuildQuorumConfig(this.qc, other_qcs, height);
         this.nominator.setQuorumConfig(this.qc, other_qcs);
-        buildRequiredKeys(this.config.validator.key_pair.address, this.qc,
-            this.required_peer_keys);
```
Comment (Geod24) [inline source/agora/node/Validator.d:144]: Why the removal ?

File: source/agora/network/Manager.d [hunk 4]
```diff
@@ -642,47 +643,60 @@ public class NetworkManager
...
+                        },
+                        )(taskman, 3, 2.seconds, "Exception happened while trying to get validator addresses");
+                    });
+                }
             }
         }
+        catch (Exception ex)
+        {
+            log.error("Error discovering peers: {}", ex);
+        }
```
Comment (Geod24) [inline source/agora/network/Manager.d]: Looks like the `try` clause is only because of the `foreach` ? If so, `byKeyValue`.

**Review by Geod24 — CHANGES_REQUESTED** [review]

File: source/agora/node/Validator.d (hunk 2 above)
Comment (linked0) [reply source/agora/node/Validator.d]: Right, I will reinstate it.

File: source/agora/network/Manager.d (hunk 4 above)
Comment (linked0) [reply source/agora/network/Manager.d]: Will update.

File: source/agora/node/Validator.d (hunk 3 above)
Comment (linked0) [reply source/agora/node/Validator.d:144]: Ditto

**Comment by linked0** [comment]
Ready for review.
I reinstated the `buildRequiredKeys` but it needs `utxo_keys` as  `rebuildQuorumConfig` does. So I hand over the variable as a parameter into the two functions in order to remove redundant calls of the functions of the `EM`.

@Geod24 What do you think I should do as the next step for solving this issue?

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
> @Geod24 What do you think I should do as the next step for solving this issue?

Unfortunately, I don't think we can switch the registry just yet, so the next step is probably to switch SCP (I've been doing some changes for that purpose), so we can at least refer to validators only by their UTXO hash in Agora (and remove the check that a validator is not already enrolled with a public key).

**Pull request merged**